controllers/terraform_controller.go

```diff
@@ -1605,3 +1583,137 @@ func (r *TerraformReconciler) templateParse(content map[string]string, text stri
 	}
 	return b.String(), nil
 }
+
+// reconcileRunnerSecret reconciles the runner secret used for mtls
+// it should create the secret if it doesn't exist and then verify that the cert is valid
+// if the cert is not present in the secret or is invalid, it will generate a new cert and
+// write it to the secret
+func (r *TerraformReconciler) reconcileRunnerSecret(ctx context.Context, terraform *infrav1.Terraform) (string, error) {
+	secretName := fmt.Sprintf("%s-runner.tls", terraform.Name)
+
+	tlsSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: terraform.Namespace,
+			Name:      secretName,
+		},
+	}
+
+	if err := r.Get(ctx, client.ObjectKeyFromObject(tlsSecret), tlsSecret); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return "", err
+		}
+		// create the runner tls secret
+		if err := r.Create(ctx, tlsSecret); err != nil {
+			return "", err
+		}
+	}
+
+	runnerHostname := fmt.Sprintf("%s.%s.svc.cluster.local", terraform.Name, terraform.Namespace)
+
+	if os.Getenv("INSECURE_LOCAL_RUNNER") == "1" {
+		runnerHostname = "localhost"
+	}
+
+	if err := r.CertRotator.RefreshRunnerCertIfNeeded(runnerHostname, tlsSecret); err != nil {
+		return "", err
+	}
+
+	return secretName, nil
+}
+
+func (r *TerraformReconciler) reconcileRunnerPod(ctx context.Context, terraform *infrav1.Terraform, tlsSecretName string) error {
+	runner := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: terraform.Namespace,
+			Name:      fmt.Sprintf("%s-runner", terraform.Name),
+		},
+	}
+	if err := r.Get(ctx, client.ObjectKeyFromObject(runner), runner); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		runner.Spec = runnerPodSpec(tlsSecretName)
+		if err := r.Create(ctx, runner); err != nil {
+			return err
+		}
+	}
+
+	runnerSvc := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: terraform.Namespace,
+			Name:      fmt.Sprintf("%s-runner", terraform.Name),
+		},
+	}
+
+	if err := r.Get(ctx, client.ObjectKeyFromObject(runnerSvc), runnerSvc); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		runnerSvc.Spec = corev1.ServiceSpec{
```

Comment:
@phoban01 Could you avoid using Service objects, and trying to connect directly to the runner pod via the pod FQDN first, please? Having a Service defined for every runner Pod will be in the plan B, if we cannot make use of Pod FQDN.


The FQDN for a pod could be obtained by prefixing its .status.PodIP with namespace and .pod.cluster.local (or the cluster domain suffix)

Here's the doc:
https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pods

Reply:
Hmm, @chanwit  thinking about this a bit more: we want mTLS to secure pod-pod communication in which case we need either a pod-ip or pod-hostname to issue the cert (otherwise we end up sharing certs). But the pod-ip isn't assigned until the pod is scheduled at which point it is too late to issue the cert (as it needs to be mounted in the pod at startup time). 

We have two options as far as I can see:
1) Create a headless service in the namespace, and specify the hostname and subdomain in the pod spec. This will allow us to connect to pods using the FQDN of the pod which will be `{hostname}.{subbomain}.{namespace}.svc.cluster.local`
2) Create one service per pod

I don't believe this would be any different using admission control as the pod-ip would not be assigned at that point either. Maybe there's another approach?

Reply:
This is an interesting problem. Could a SAN wildcard `*.{namespace}.pod.cluster.local` solve this?

Reply:
Yes that would work, with the proviso that mTLS is scoped at the namespace level rather than per-pod.

Reply:
That's great. Thank you, @phoban01.

Reply:
Have made a change to address this @chanwit.